src/util.rs

```diff
@@ -355,3 +356,33 @@ fn get_network_interfaces_impl() -> Vec<NetworkInterface> {
     vec
 }
 
+/// checksum computes a TCP or UDP checksum.
+/// raw_packet is the TCP or UDP packet.
+/// encapsulating_packet is either the ipv4 or ipv6 packet encapsulating the higher protocol.
+pub fn checksum<T: PseudoHeader>(raw_packet: &[u8], encapsulating_packet: T) -> u16 {
+    let mut sum = encapsulating_packet.checksum();
+    let length = raw_packet.len() as u32;
+    sum = sum + length & 0xffff;
+    sum = sum + length >> 16;
+    return rfc1071_checksum(raw_packet, sum);
+}
+
+/// Calculates rfc1071 checksum value
+pub fn rfc1071_checksum(packet: &[u8], initial: u32) -> u16 {
+    let length = packet.len() - 1;
+    let mut sum = initial;
+
+    let mut i = 0;
+    while i < length {
+        let word = (packet[i] as u32) << 8 | packet[i + 1] as u32;
+        sum = sum + word;
+        i = i + 2;
+    }
+    if packet.len()%2 == 1 {
+        sum = sum + (packet[length] as u32) << 8
```

Comment:
@david415 : I _think_ this is equivalent to `sum = (sum + (packet[length] as u32)) << 8` (at least that's what `clippy` suggested to me), whereas we want to do `sum = sum + ((packet[length] as u32) << 8)`:

```
src/util.rs:399:15: 399:49 warning: operator precedence can trip the unwary. Consider parenthesizing your expression:`(sum + (packet[length] as u32)) << 8`, #[warn(precedence)] on by default
src/util.rs:399         sum = sum + (packet[length] as u32) << 8
```

Also, `;` is missing, so the result is returned before the carries are added.

So changing this line `sum = sum + ((packet[length] as u32) << 8);` makes the following test pass:

```rust
#[test]
fn checksum_odd() {
    let data = vec![
        0x03, 0x01, 0x00, 0x00, 0x12, 0x34, 0x56, 0x78,
        0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
        0x55];
    //   03 01
    // + 00 00
    // + 12 34
    // + 56 78
    // + 01 02
    // + 03 04
    // + 05 06
    // + 07 08
    // + 55 00
    //  +2 +2
    // --------
    //   d0 c1 = 1101 0000 1100 0001
    //
    // => checksum is 0010 1111 0011 1110
    //                   2    f    3    e
    assert_eq!(rfc1071_checksum(&data[..], 0), 0x2f3e);
}
```